Exposing tmsis command

kkroo opened the pull request:
Allows you to make calls to `tmsis`, a command that already exists in openbts's nodemanager interface

kkroo committed:
Exposing tmsis command

yosemitebandit commented on openbts/tests/openbts_component_tests.py:
```diff
@@ -154,6 +154,38 @@ def test_get_version(self):
+    self.openbts_connection.socket = mock.Mock()
+    self.openbts_connection.socket.recv.return_value = json.dumps({
+      'code': 200,
+      'data': {
```
does it return a list of dicts or just a single dict?  (not sure, just going on the comment in the `tmsis` method above)

kkroo replied:
Yah this should be a list of dicts.. updating this

yosemitebandit commented on openbts/components.py:
```diff
@@ -41,6 +41,40 @@ def monitor(self):
+      'action': '',
+      'key': '',
+      'value': ''
+    }
```
what do you think about doing some processing here, like returning a dict that is keyed by IMSI?

kkroo replied:
i'm thinking of allowing some optional filtering on `ACCESS` or `AUTH` but I'm not quite sure how useful a lookup by IMSI would be

yosemitebandit replied:
I was imagining how `client` would use this API.. and it seems like `client` would call want to report the `ACCESSED` time or something for each IMSI, so I thought a large dict with a bunch of data for each IMSI, keyed by IMSI, would be easiest.

yosemitebandit commented:
Can you add a barebones integration test for this?  That module kind of encapsulates all of the functionality of the package.

shaddi commented:
In line with the change in endaga/openbts#19, for our purposes we only care about the # of seconds since last accessed, rather than the timestamp of accessed/created itself. Should we consider adding this time to the result here? I am not sure whether it makes more sense to add that value here in postprocessing, or have OpenBTS provide that itself, or leave that responsibility to the caller. 

My concern here comes from the fact that in other NM responses that contain timestamps I've seen the timestamp just be very incorrect (PhysicalStatus comes to mind), so we should at least test that to make sure we _can_ get the right value from the caller if we're going down that route.

kkroo commented:
Good point, we could just leave OpenBTS to handle computing time differences rather than giving the epoch in endaga/openbts#19

yosemitebandit commented on openbts/components.py:
```diff
@@ -41,6 +41,40 @@ def monitor(self):
+    """Gets all active subscribers from the TMSI table
+
+    Will return a list of objects of the form: {
+            'A5_SUPPORT' : '',
```
more nits: can you explain some / all of these keys in comments?  Or link to docs?

kkroo committed:
Using the JSONDB access method for more efficient filter

kkroo committed:
adding tests

kkroo commented:
So I changed the way that this is pulling the TMSIs (See endaga/openbts#23) but now we can at least filter out subscribers who are not on our network in SQL. This should save us from having to send over the entire TMSI table over zmq on checkin.

Added integration test as requested @yosemitebandit, but it is a pretty silly one for now

PTAL

shaddi commented:
We should update the OpenBTS depends on this one if it's depending on new NodeManager API behavior...

kkroo commented:
Yeah I forgot the version bump

kkroo committed:
bumping version to 0.1.1

yosemitebandit commented on openbts/components.py:
```diff
@@ -41,6 +42,37 @@ def monitor(self):
+  def tmsis(self, fields=['IMSI', 'ACCESSED'], access_period=0, auth=2):
+    """Gets all active subscribers from the TMSI table.
+
+    Args: access_period: fetches all entries with ACCESS < acess_period
```
nit: typo in second occurrence of access_period

yosemitebandit commented on openbts/components.py:
```diff
@@ -41,6 +42,37 @@ def monitor(self):
+
+    Args: access_period: fetches all entries with ACCESS < acess_period
+                         (default=0 filter off)
+          auth: fetches all entries with AUTH = auth
```
what are these auth levels?

kkroo replied:
They are defined by the [Authorization enum](https://github.com/endaga/openbts/blob/master/Control/TMSITable.h#L45) in the TMSI table. This field tells us whether a MS is authorized to use the network.The reason for allowing filtering on AUTH is that the TMSI table can actually get quite full with non authorized users, so this field allows developers to fetch the subscribers they care about. I should probably update the docstring to list out the different values since they are not listed in OpenBTS docs.

yosemitebandit commented on openbts/components.py:
```diff
@@ -41,6 +42,37 @@ def monitor(self):
+          auth: fetches all entries with AUTH = auth
+                         (default=2)
+
+    Returns: a list of objects defined by the list of fields.
```
I feel like this API should help out the user more -- like there should be a consistent reply and you shouldn't have to specify fields.  Why not just return IMEI, IMSI, TMSI, CREATED, ACCESSED and other fields that you think we'll need in `client`?

I guess I'm suggesting you make this method more than a light wrapper around NM.  I think if this command does more processing / parsing, the people using it in `client` will have an easier time..

yosemitebandit replied:
and I see the benefits of making a more generic command, but I think it makes it harder to understand how to use these methods.  We can add fields and do another release later if need be.

yosemitebandit replied:
sorry to harangue on this, just feel like it'll help us out in `client` (:

kkroo replied:
I see your point with this one, better to keep it generic and take what we need on the client side

kkroo committed:
extending fields returned by tmsis and doc cleanup

yosemitebandit commented:
is it true that this will only work on our fork of OpenBTS?  If so, can you add a comment to that effect?

yosemitebandit commented:
I take it this is ready for review too?   Sorry this has been so delayed :(

yosemitebandit commented on integration_tests.py:
```diff
@@ -86,6 +86,15 @@ def test_monitor_openbts(self):
     response = connection.monitor()
     self.assertIn('noiseRSSI', response.data)
 
```
ubernit: two blank lines here and below

kkroo commented:
Yes, PTAL :)

yosemitebandit commented:
LGTM, LMK when I should push to pypi

kkroo commented:
We can push this to pypi once endaga/openbts#23 is live.
Sry bout the chain dependencies!

kkroo committed:
whitespace fix

kkroo committed:
readme fix

kkroo commented:
Alrighty, this is ready for pypi!

yosemitebandit commented:
alrighty, 0.1.1 is released to pypi
